File: models/product.py
# -*- coding: utf-8 -*-
"""
models/product.py
จัดการข้อมูลสินค้า: CRUD, ราคาหลายช่องทางขาย, และการปรับเพิ่ม/ลดสต็อก
"""

import uuid
from datetime import datetime

import config
from database.db import get_connection


def _now():
    return datetime.now()


def _new_id():
    return str(uuid.uuid4())


def _row_to_dict_with_prices(conn, row):
    product = dict(row)
    cur = conn.cursor()
    cur.execute(
        "SELECT channel, price, discount_type, discount_value FROM product_price WHERE product_id = %s",
        (product["id"],),
    )
    product["prices"] = {
        r["channel"]: {
            "price": r["price"],
            "discount": r["discount_value"] if r["discount_type"] != "NONE" else 0,
        }
        for r in cur.fetchall()
    }

    cur.execute("SELECT name FROM category WHERE id = %s", (product["category_id"],))
    cat_row = cur.fetchone()
    product["category_name"] = cat_row["name"] if cat_row else "ไม่ระบุหมวดหมู่"
    return product


def get_all_products(search_text: str = None):
    """ดึงสินค้าที่ยัง active อยู่ทั้งหมด พร้อมรองรับค้นหาจาก SKU หรือชื่อสินค้า"""
    conn = get_connection()
    try:
        cur = conn.cursor()
        if search_text:
            like = f"%{search_text.strip()}%"
            cur.execute(
                "SELECT * FROM product WHERE is_active = TRUE AND (sku LIKE %s OR name LIKE %s) ORDER BY name",
                (like, like),
            )
        else:
            cur.execute("SELECT * FROM product WHERE is_active = TRUE ORDER BY name")
        rows = cur.fetchall()
        return [_row_to_dict_with_prices(conn, row) for row in rows]
    finally:
        conn.close()


def get_product_by_id(product_id):
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute("SELECT * FROM product WHERE id = %s", (product_id,))
        row = cur.fetchone()
        if row is None:
            return None
        return _row_to_dict_with_prices(conn, row)
    finally:
        conn.close()


def get_low_stock_products():
    """สินค้าที่จำนวนคงเหลือน้อยกว่า min_stock_alert ของตัวเอง"""
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute(
            "SELECT * FROM product WHERE is_active = TRUE AND current_stock < min_stock_alert ORDER BY current_stock ASC"
        )
        rows = cur.fetchall()
        return [_row_to_dict_with_prices(conn, row) for row in rows]
    finally:
        conn.close()


def get_total_stock_summary():
    """สรุปยอดรวมสำหรับหน้า Dashboard: จำนวนสินค้าทั้งหมด และจำนวนรวมของสต็อกทุกชิ้น"""
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute(
            "SELECT COUNT(*) as total_products, COALESCE(SUM(current_stock), 0) as total_quantity "
            "FROM product WHERE is_active = TRUE"
        )
        row = cur.fetchone()
        return {"total_products": row["total_products"], "total_quantity": row["total_quantity"]}
    finally:
        conn.close()


def create_product(data: dict, user_id: str) -> str:
    """
    สร้างสินค้าใหม่ คืนค่า id ที่สร้าง (string)
    data ต้องมี: code(sku), name,common_name, category_id, base_unit, pack_unit, units_per_pack,
                quantity, low_stock_threshold, image_url, note, prices (dict: {channel: {price, discount}})
    """
    conn = get_connection()
    try:
        cur = conn.cursor()
        now = _now()
        product_id = _new_id()
        cur.execute(
            """
            INSERT INTO product
                (id, sku, name,common_name, category_id, base_unit, pack_unit, units_per_pack, current_stock,
                 min_stock_alert, image_url, description, is_active, created_by_id, updated_by_id, created_at, updated_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """,
            (
                product_id,
                data["code"].strip() or None,
                data["name"].strip(),
                data["common_name"].strip(),
                data.get("category_id"),
                data["packaging_unit"].strip(),
                data.get("pack_unit") or None,
                int(data.get("units_per_pack", 1)),
                int(data.get("quantity", 0)),
                int(data.get("low_stock_threshold", config.LOW_STOCK_DEFAULT_THRESHOLD)),
                data.get("image_path") or None,
                data.get("note") or None,
                True,
                user_id,
                user_id,
                now,
                now,
            ),
        )

        for channel, price_info in (data.get("prices") or {}).items():
            discount_value = float(price_info.get("discount", 0) or 0)
            cur.execute(
                """
                INSERT INTO product_price (id, product_id, channel, price, discount_type, discount_value, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s)
                """,
                (
                    _new_id(),
                    product_id,
                    channel,
                    float(price_info.get("price", 0)),
                    "FIXED" if discount_value else "NONE",
                    discount_value,
                    now,
                ),
            )

        conn.commit()
        return product_id
    finally:
        conn.close()


def update_product(product_id, data: dict, user_id: str):
    """แก้ไขข้อมูลสินค้าที่มีอยู่ (ไม่รวมจำนวนสต็อก ซึ่งต้องปรับผ่าน adjust_stock เพื่อให้มี log ที่ถูกต้อง)"""
    conn = get_connection()
    try:
        cur = conn.cursor()
        now = _now()
        cur.execute(
            """
            UPDATE product
            SET sku = %s, name = %s,common_name = %s, category_id = %s, base_unit = %s, pack_unit = %s, units_per_pack = %s,
                min_stock_alert = %s, image_url = %s, description = %s, updated_by_id = %s, updated_at = %s
            WHERE id = %s
            """,
            (
                data["code"].strip() or None,
                data["name"].strip(),
                data["common_name"].strip(),
                data.get("category_id"),
                data["packaging_unit"].strip(),
                data.get("pack_unit") or None,
                int(data.get("units_per_pack", 1)),
                int(data.get("low_stock_threshold", config.LOW_STOCK_DEFAULT_THRESHOLD)),
                data.get("image_path") or None,
                data.get("note") or None,
                user_id,
                now,
                product_id,
            ),
        )

        # อัปเดตราคา: ลบของเก่าแล้วใส่ใหม่ทั้งหมด ง่ายและชัดเจนกว่าการไล่ diff ทีละช่องทาง
        cur.execute("DELETE FROM product_price WHERE product_id = %s", (product_id,))
        for channel, price_info in (data.get("prices") or {}).items():
            discount_value = float(price_info.get("discount", 0) or 0)
            cur.execute(
                """
                INSERT INTO product_price (id, product_id, channel, price, discount_type, discount_value, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s)
                """,
                (
                    _new_id(),
                    product_id,
                    channel,
                    float(price_info.get("price", 0)),
                    "FIXED" if discount_value else "NONE",
                    discount_value,
                    now,
                ),
            )

        conn.commit()
    finally:
        conn.close()


def deactivate_product(product_id):
    """ลบสินค้าแบบ soft-delete (is_active = False) เก็บประวัติ stock_movement/activity_log ไว้ครบ"""
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute("UPDATE product SET is_active = FALSE WHERE id = %s", (product_id,))
        conn.commit()
    finally:
        conn.close()


def adjust_stock(product_id, change_amount: int, reason: str, unit_label: str, input_quantity: float, user_id: str):
    """
    เพิ่ม/ลดจำนวนสต็อกสินค้า (change_amount เป็นบวก = เพิ่ม (IN), ลบ = ลด (OUT), หน่วยเป็นหน่วยย่อยเสมอ)
    unit_label/input_quantity: หน่วยและจำนวนตามที่ผู้ใช้กรอกจริง (ก่อนแปลงเป็นหน่วยย่อย) เก็บไว้เป็นประวัติ
    บันทึกลง stock_movement ด้วยทุกครั้งเพื่อดูประวัติย้อนหลังได้
    """
    conn = get_connection()
    try:
        cur = conn.cursor()
        cur.execute("SELECT current_stock FROM product WHERE id = %s", (product_id,))
        row = cur.fetchone()
        if row is None:
            raise ValueError("ไม่พบสินค้านี้ในระบบ")

        new_quantity = row["current_stock"] + change_amount
        if new_quantity < 0:
            raise ValueError("จำนวนสต็อกคงเหลือจะติดลบ กรุณาตรวจสอบจำนวนที่ต้องการลด")

        cur.execute("UPDATE product SET current_stock = %s, updated_by_id = %s WHERE id = %s", (new_quantity, user_id, product_id))
        cur.execute(
            """
            INSERT INTO stock_movement
                (id, product_id, type, quantity, input_unit, input_quantity, balance_before, balance_after,
                 note, created_by_id, created_at)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """,
            (
                _new_id(),
                product_id,
                "IN" if change_amount > 0 else "OUT",
                abs(change_amount),
                unit_label,
                input_quantity,
                row["current_stock"],
                new_quantity,
                reason or None,
                user_id,
                _now(),
            ),
        )
        conn.commit()
        return new_quantity
    finally:
        conn.close()

File: website/models/Product.ts
import { Prisma } from "@prisma/client";

import { prisma } from "@/lib/prisma";
import { newId } from "@/lib/id";
import { AppError } from "@/lib/errors";
import { LOW_STOCK_DEFAULT_THRESHOLD } from "@/lib/constants";

const productInclude = { prices: true, category: true } satisfies Prisma.ProductInclude;
type ProductWithRelations = Prisma.ProductGetPayload<{ include: typeof productInclude }>;

export type ProductPriceInput = { price: number; discount?: number };

export type ProductInput = {
  sku: string;
  name: string;
  commonName?: string;
  categoryId: string;
  baseUnit: string;
  packUnit?: string;
  unitsPerPack: number;
  minStockAlert: number;
  imageUrl?: string;
  description?: string;
  prices: Record<string, ProductPriceInput>;
};

/** คืน error ถ้าข้อมูลไม่ถูกต้อง หรือคืน void ถ้าผ่าน (โยน AppError เหมือน validate_product_data ฝั่ง desktop) */
export function validateProductInput(data: Partial<ProductInput> & { quantity?: number }) {
  if (!data.sku?.toString().trim()) throw new AppError("กรุณากรอกรหัสสินค้า");
  if (!data.name?.toString().trim()) throw new AppError("กรุณากรอกชื่อสินค้า");
  if (!data.baseUnit?.toString().trim()) throw new AppError("กรุณากรอกหน่วยนับย่อย (เช่น ขวด, ซอง, ชิ้น)");

  const unitsPerPack = Number(data.unitsPerPack ?? 1);
  if (!Number.isFinite(unitsPerPack) || unitsPerPack <= 0) {
    throw new AppError("จำนวนหน่วยย่อยต่อแพ็คต้องมากกว่า 0");
  }

  if (data.quantity !== undefined) {
    const quantity = Number(data.quantity);
    if (!Number.isFinite(quantity) || quantity < 0) throw new AppError("จำนวนสินค้าต้องไม่ติดลบ");
  }

  const minStockAlert = Number(data.minStockAlert ?? LOW_STOCK_DEFAULT_THRESHOLD);
  if (!Number.isFinite(minStockAlert) || minStockAlert < 0) {
    throw new AppError("ค่าสต็อกขั้นต่ำต้องไม่ติดลบ");
  }
}

function toProductDTO(product: ProductWithRelations) {
  const { prices, category, ...rest } = product;
  return {
    ...rest,
    categoryName: category?.name ?? "ไม่ระบุหมวดหมู่",
    prices: Object.fromEntries(
      prices.map((p) => [p.channel, { price: p.price, discount: p.discountType !== "NONE" ? p.discountValue : 0 }])
    ),
  };
}

export async function getAllProducts(search?: string) {
  const keyword = search?.trim();
  const products = await prisma.product.findMany({
    where: {
      isActive: true,
      ...(keyword ? { OR: [{ sku: { contains: keyword } }, { name: { contains: keyword } }] } : {}),
    },
    include: productInclude,
    orderBy: { name: "asc" },
  });
  return products.map(toProductDTO);
}

export async function getProductById(id: string) {
  const product = await prisma.product.findUnique({
    where: { id },
    include: productInclude,
  });
  return product ? toProductDTO(product) : null;
}

/** สินค้าที่จำนวนคงเหลือน้อยกว่า min_stock_alert ของตัวเอง (เปรียบเทียบ 2 คอลัมน์ ต้องใช้ raw query) */
export async function getLowStockProducts() {
  const rows = await prisma.$queryRaw<{ id: string }[]>`
    SELECT id FROM product WHERE is_active = 1 AND current_stock < min_stock_alert ORDER BY current_stock ASC
  `;
  if (rows.length === 0) return [];

  const products = await prisma.product.findMany({
    where: { id: { in: rows.map((r) => r.id) } },
    include: productInclude,
  });
  const byId = new Map(products.map((p) => [p.id, p]));
  return rows.map((r) => toProductDTO(byId.get(r.id)!));
}

export async function getTotalStockSummary() {
  const result = await prisma.product.aggregate({
    where: { isActive: true },
    _count: { _all: true },
    _sum: { currentStock: true },
  });
  return { totalProducts: result._count._all, totalQuantity: result._sum.currentStock ?? 0 };
}

function priceCreateData(now: Date, channel: string, info: ProductPriceInput) {
  const discountValue = info.discount ?? 0;
  return {
    id: newId(),
    channel,
    price: info.price,
    discountType: discountValue ? "FIXED" : "NONE",
    discountValue,
    updatedAt: now,
  };
}

export async function createProduct(data: ProductInput & { quantity: number }, userId: string) {
  const now = new Date();
  const product = await prisma.product.create({
    data: {
      id: newId(),
      sku: data.sku.trim() || null,
      name: data.name.trim(),
      commonName: data.commonName?.trim() || null,
      categoryId: data.categoryId,
      baseUnit: data.baseUnit.trim(),
      packUnit: data.packUnit?.trim() || null,
      unitsPerPack: Math.trunc(data.unitsPerPack),
      currentStock: Math.trunc(data.quantity),
      minStockAlert: Math.trunc(data.minStockAlert),
      imageUrl: data.imageUrl?.trim() || null,
      description: data.description?.trim() || null,
      isActive: true,
      createdById: userId,
      updatedById: userId,
      updatedAt: now,
      prices: {
        create: Object.entries(data.prices ?? {}).map(([channel, info]) => priceCreateData(now, channel, info)),
      },
    },
  });
  return product.id;
}

/** แก้ไขข้อมูลสินค้าที่มีอยู่ (ไม่รวมจำนวนสต็อก ต้องปรับผ่าน adjustStock เพื่อให้มี log ที่ถูกต้อง) */
export async function updateProduct(id: string, data: ProductInput, userId: string) {
  const now = new Date();
  await prisma.$transaction([
    prisma.product.update({
      where: { id },
      data: {
        sku: data.sku.trim() || null,
        name: data.name.trim(),
        commonName: data.commonName?.trim() || null,
        categoryId: data.categoryId,
        baseUnit: data.baseUnit.trim(),
        packUnit: data.packUnit?.trim() || null,
        unitsPerPack: Math.trunc(data.unitsPerPack),
        minStockAlert: Math.trunc(data.minStockAlert),
        imageUrl: data.imageUrl?.trim() || null,
        description: data.description?.trim() || null,
        updatedById: userId,
        updatedAt: now,
      },
    }),
    prisma.productPrice.deleteMany({ where: { productId: id } }),
    ...Object.entries(data.prices ?? {}).map(([channel, info]) =>
      prisma.productPrice.create({ data: { ...priceCreateData(now, channel, info), productId: id } })
    ),
  ]);
}

/** ลบสินค้าแบบ soft-delete (is_active = false) เก็บประวัติ stock_movement/activity_log ไว้ครบ */
export async function deactivateProduct(id: string) {
  await prisma.product.update({ where: { id }, data: { isActive: false } });
}
